### PR TITLE
Add on_error parameter to AuthenticationMiddleware

### PR DESCRIPTION
**Pull request by pvanliefland**
(Replaces PR #278)

Usage:

```python
def on_auth_error(request: Request, exc: Exception):
    return JSONResponse({"error": str(exc)}, status_code=401)

app.add_middleware(AuthenticationMiddleware, backend=BasicAuthBackend(), on_error=on_auth_error)
```

### CODE REVIEW

File: starlette/middleware/authentication.py [hunk 1]
```diff
@@ -11,10 +12,20 @@
 from starlette.types import ASGIApp, ASGIInstance, Receive, Scope, Send
 
 
+def default_on_error(request: Request, exc: Exception):
```
Comment (lovelydinosaur) [inline starlette/middleware/authentication.py]: `-> Response`

File: starlette/middleware/authentication.py [hunk 2]
```diff
@@ -11,10 +12,20 @@
...
-    def __init__(self, app: ASGIApp, backend: AuthenticationBackend) -> None:
+    def __init__(
+        self,
+        app: ASGIApp,
+        backend: AuthenticationBackend,
+        on_error: typing.Callable = None,
+    ) -> None:
         self.app = app
         self.backend = backend
+        self.on_error = on_error if on_error is not None else default_on_error
```
Comment (lovelydinosaur) [inline starlette/middleware/authentication.py]: Could we have the default handler be a method on the class?
`self.on_error = on_error if on_error is not None else self.default_on_error`

File: starlette/middleware/authentication.py [hunk 3]
```diff
@@ -11,10 +12,20 @@
...
+    return PlainTextResponse(str(exc), status_code=400)
+
+
 class AuthenticationMiddleware:
-    def __init__(self, app: ASGIApp, backend: AuthenticationBackend) -> None:
+    def __init__(
+        self,
+        app: ASGIApp,
+        backend: AuthenticationBackend,
+        on_error: typing.Callable = None,
```
Comment (lovelydinosaur) [inline starlette/middleware/authentication.py]: Might need a more complete annotation here? `typing.Callable[[Request, AuthenticationError], Response]` perhaps?

**Comment by pvanliefland** [comment]
Cool :)

**Pull request merged**